src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,464 @@
+        }
+
+        /// <summary>
+        /// Returns a text span with trailing whitespace trimmed.
+        /// </summary>
+        public static ReadOnlyMemory<char> TrimEndWhiteSpace(ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+                return memory;
+
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            if (!char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                return memory;
+
+            while (ichMin < ichLim && char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                ichLim--;
+
+            return outerBuffer.AsMemory().Slice(ichMin, ichLim - ichMin);
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Single value, ReadOnlyMemory<char> memory)
+        {
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            var res = DoubleParser.Parse(out value, outerBuffer, ichMin, ichLim);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Double value, ReadOnlyMemory<char> memory)
+        {
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            var res = DoubleParser.Parse(out value, outerBuffer, ichMin, ichLim);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        public static uint Hash(uint seed, ReadOnlyMemory<char> memory)
+        {
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            return Hashing.MurmurHash(seed, outerBuffer, ichMin, ichLim);
+        }
+
+        // REVIEW: Add method to NormStr.Pool that deal with ReadOnlyMemory instead of the other way around.
+        public static NormStr AddToPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            return pool.Add(outerBuffer, ichMin, ichLim);
+        }
+
+        public static NormStr FindInPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
```

Comment:
Same comment as before. You're either supporting `ReadOnlyMemor<char>` or you're not. Also of course as with every other usage of this function we are ignoring whether this succeeded or not.